### PR TITLE
Issue-1154: Add Support for Handling Deleted Articles

### LINKED ISSUE #1154: Add Support for Handling Deleted Articles
### Description

I would like to see support for removing the Apple News article metadata in WP if it tries to get info on the articles from the News API and the News API reports that the article doesn't exist, so that error messages are no longer thrown for this scenario.

### Use Case

When a post is deleted directly in iCloud News Publisher, the postmeta should be updated in WP if the News API reports a 404 so the request to fetch the article doesn't continue.

# Acceptance Criteria

- When fetching information about an article using the News API, if the News API reports that an article does not exist (response code [404](https://developer.apple.com/documentation/apple_news/read_article_information)) delete the Publish to Apple News postmeta that tracks the article details (e.g., API ID, last updated date, etc). Do not do this on other types of bad responses (e.g., 400, 401, 403).
- Add a unit test to confirm this behavior.

### PR DESCRIPTION
**Pull request by renatonascalves**
### Summary

This PR adds support for handling scenarios where the Apple News API reports that an article does not exist anymore, by removing the associated metadata in WordPress. This aims to prevent error messages from being thrown when information on deleted articles is requested. Fixes #1154

### Description

Implemented a check to remove the "Publish to Apple News" postmeta data in WordPress when the News API returns a 404 error code, indicating that the article does not exist. This ensures that WordPress does not continue to attempt fetching data for articles that have been deleted from Apple News.

### Use Case

This functionality is particularly useful when a post is deleted directly in iCloud News Publisher but still exists in WordPress. With this update, the postmeta in WordPress will be automatically updated to reflect the deletion, based on the News API's response, thus avoiding unnecessary fetch requests and error messages.

### Acceptance Criteria

- When the News API reports a 404 error for an article, the associated "Publish to Apple News" postmeta (including API ID, last updated date, etc.) is deleted from WordPress.
- This behavior is confirmed through the addition of a unit test.

### Testing Instructions

1. Ensure that the unit test for the new behavior passes.
2. Manually test by simulating a 404 response from the News API for an article and verify that the corresponding postmeta is deleted in WordPress.

### CODE REVIEW

File: includes/REST/apple-news-get-published-state.php [hunk 1]
```diff
@@ -7,40 +7,60 @@
...
+ */
+function get_published_state_response( $request ): WP_REST_Response|WP_Error {
+	$id = $request->get_param( 'id' );
+
+	// Ensure Apple News is first initialized.
+	$retval = \Apple_News::has_uninitialized_error();
+
+	if ( is_wp_error( $retval ) ) {
+		return $retval;
+	}
```
Comment (renatonascalves) [inline includes/REST/apple-news-get-published-state.php:57]: I'm currently only fixing this in this endpoint, since I needed to test the new behavior for deleted articles. 

A separate issue will handle it for the other endpoints.

File: admin/apple-actions/index/class-get.php [hunk 2]
```diff
@@ -43,18 +41,28 @@ public function __construct( $settings, $id ) {
...
+		try {
+			$article = $this->get_api()->get_article( $apple_id );
+		} catch ( \Apple_Push_API\Request\Request_Exception $e ) {
+			$article = $e->getMessage();
+		}
+
+		// Reset the API postmeta if the article is not present in Apple News.
+		if ( is_string( $article ) && str_contains( $article, 'NOT_FOUND (keyPath articleId)' ) ) {
+			$this->reset_meta( $this->id );
+		}
```
Comment (renatonascalves) [inline admin/apple-actions/index/class-get.php]: To avoid a bigger refactoring of the get_api implemenation, I decided to check against this here. I confirmed in Apple News [documentation](https://developer.apple.com/documentation/apple_news/read_article_information) that `NOT_FOUND` is only sent when an article 404s.

Comment (renatonascalves) [reply admin/apple-actions/index/class-get.php]: So it seems like a good place to reset the meta from the post.

File: admin/apple-actions/class-api-action.php [hunk 3]
```diff
@@ -88,4 +88,17 @@ protected function is_api_configuration_valid() {
 
 		return true;
 	}
+
+	/**
+	 * Resets the API postmeta for a given post ID.
+	 *
+	 * @param int $post_id The post ID to reset.
+	 */
+	protected function reset_meta( $post_id ): void {
```
Comment (kingkool68) [inline admin/apple-actions/class-api-action.php]: `delete_post_meta()` comes to mind first for the name of this method name

File: admin/apple-actions/index/class-get.php (hunk 2 above)
Comment (kingkool68) [reply admin/apple-actions/index/class-get.php]: Wouldn't you only want to run this if an error was detected from the `catch ( \Apple_Push_API\Request\Request_Exception $e ) {` statement above?

**Review by kingkool68 — APPROVED** [review]
👍

File: admin/apple-actions/index/class-get.php (hunk 2 above)
Comment (renatonascalves) [reply admin/apple-actions/index/class-get.php]: Might be safer. Added.

File: admin/apple-actions/class-api-action.php (hunk 3 above)
Comment (renatonascalves) [reply admin/apple-actions/class-api-action.php]: Updated!

**Comment by renatonascalves** [comment]
This pr is approved but currently blocked pending some testing on a live/testing site (tested only locally). Please, do not merge before it.

**Comment by kingkool68** [comment]
**tl;dr: After testing this out locally the only thing that didn't happen was seeing any sort of error message after updating the post after deleting the Apple News article via iCloud.**

Testing this out locally using the following steps:
 - [x] Publish an article and confirm it is live in Apple News.
 - [x] Log in to iCloud Apple News and delete the article in Apple News (not WordPress).
 - [ ] Try to update the article, any update will do, and confirm you see this error message
   - I did not see any sort of error message after updating the article  
 - [x] Click save and reload the page.
 - [x] Confirm any apple_news related meta keys are removed.
 - [x] Confirm saving the post resends/reindexes the post in Apple News (with a new articleId)

Post meta after my article was published on Apple News:
Screenshot 2024-09-19 at 11 05 38 AM

Post meta after the article was deleted via icloud and then updated in WordPress:
Screenshot 2024-09-19 at 11 10 55 AM

Once the article is deleted via iCloud any subsequent updates in WordPress aren't reflected in Apple News. The article is never re-published on Apple News again. I think this is the desired behavior here but I can see how this would be confusing to a normal editor. Viewing the Apple News Index in the WordPress admin (Apple news --> Articles) shows the Apple News status as LIVE and Sync Status as "Needs to be updated"

Screenshot 2024-09-19 at 11 30 21 AM

Environment:
 - WordPress 6.6.2
 -  The Apple News plugin from this branch uploaded to WordPress as a zip file

**Comment by renatonascalves** [comment]
I tested this locally again and it seems to be working as expected.

Post meta after my article was published on Apple News:

CleanShot 2024-10-02 at 17 45 00@2x

Post meta **after** the article was deleted via icloud and then updated in WordPress:
CleanShot 2024-10-02 at 17 45 50@2x

I'll merge this pr and test in a staging environment. I need to merge it first to take advantage of the [built branch](https://github.com/alleyinteractive/apple-news/tree/develop-built).
